src/facade/reply_builder_test.cc

```diff
@@ -0,0 +1,706 @@
+      return args.size() == 1 && results == RedisParser::OK && args[0].type == RespExpr::STRING;
+    }
+  };
+
+  static void SetUpTestSuite() {
+  }
+
+  void SetUp() {
+    sink_.Clear();
+    builder_.reset(new RedisReplyBuilder(&sink_));
+  }
+
+ protected:
+  std::vector<std::string_view> RawTokenizedMessage() const {
+    CHECK(!str().empty());
+    return absl::StrSplit(str(), kCRLF);
+  }
+
+  std::string_view str() const {
+    return sink_.str();
+  }
+
+  std::string TakePayload() {
+    std::string ret = sink_.str();
+    sink_.Clear();
+    return ret;
+  }
+
+  std::size_t SinkSize() const {
+    return str().size();
+  }
+
+  // Breaks the string we have in sink into tokens.
+  // In  RESP each token is build up from series of bytes follow by "\r\n"
+  // This function don't try to parse the message, only to break the strings based
+  // on the delimiter "\r\n". It is up to the test to verify these tokens
+  std::vector<std::string_view> TokenizeMessage() const;
+
+  // Call the redis parser with the data in the sink
+  ParsingResults Parse();
+
+  io::StringSink sink_;
+  std::unique_ptr<RedisReplyBuilder> builder_;
+  std::unique_ptr<std::uint8_t[]> parser_buffer_;
+};
+
+std::vector<std::string_view> RedisReplyBuilderTest::TokenizeMessage() const {
+  std::vector<std::string_view> message_tokens = RawTokenizedMessage();
+  CHECK(message_tokens.back().empty());  // we're expecting to last to be empty as it only has \r\n
+  message_tokens.pop_back();             // remove this empty entry
+  std::string_view data = str();
+  switch (data[0]) {
+    case kArrayStart:
+      // in the case of array. we cannot tell the expected tokens number without doing parsing for
+      // sub elements
+      break;
+    case kBulkString:
+      if (data == kNullBulkString) {
+        CHECK(message_tokens.size() == 1)
+            << "NULL bulk string should only have one token, got " << message_tokens.size();
+      } else {
+        CHECK(message_tokens.size() == 2)
+            << "bulk string should only have two tokens, got " << message_tokens.size();
+      }
+      break;
+    case kErrorStartChar:
+    case kStringStartChar:
+    case kIntStart:
+      // for errors and string and ints we don't really need to split as there must be only one
+      // entry for \r\n
+      CHECK(message_tokens.size() == 1)
+          << "string/error message must have only one token got " << message_tokens.size();
+      break;
+    default:
+      CHECK(false) << "invalid start char [" << data[0] << "]";
+      break;
+  }
+  return message_tokens;
+}
+
+std::ostream& operator<<(std::ostream& os, const RedisReplyBuilderTest::ParsingResults& res) {
+  os << "result{consumed bytes:" << res.consumed << ", status: " << res.results << " result count "
+     << res.args.size() << ", first entry result: ";
+  if (!res.args.empty()) {
+    if (res.args.size() > 1) {
+      os << "ARRAY: ";
+    }
+    int index = 0;
+    for (const auto& e : res.args) {
```

Comment:
why not use `ostream& operator<<(ostream& os, const facade::RespExpr& e)` defined in facade.cc ?

Reply:
I think I missed that - I will change this. you're right